### PR TITLE
feat(codecs, sources): Implement GELF encoder

### LINKED ISSUE #4868: GELF codec support
Support GELF as a codec for use when sending data to Graylog: https://docs.graylog.org/en/3.3/pages/sending/gelf.html

GELF can be sent over HTTP, UDP, and TCP to Graylog.

We should add `gelf` as a codec option for both sources and sinks.

Example configuration:

```toml
[sources.foo]
type = "stdin"
decoding.codec = "gelf"
```

```toml
[sinks.bar]
type = "console"
encoding.codec = "gelf"
```

### PR DESCRIPTION
**Pull request by neuronull**
Closes #4868 

This is the encoding portion of "GELF codec support" https://github.com/vectordotdev/vector/issues/4868.

The decoding portion is in PR #13288 .

Some notes about the behavior of this encoder:

- GELF standard is documented here: https://docs.graylog.org/docs/gelf
- Graylog itself follows a relaxed approach to enforcing their own standard.
- In order to provide a backwards compatible product, we've elected to take a strict approach to handling GELF decoding and encoding.
- For encoding specifically, we adhere strictly to the standard with the following exception:
    - According to the spec, 'Additional fields' must be prefixed with an underscore (eg "_foo"). We've elected to prepend that character to any (Additional) field that does not have it. This is because it does not detract much from the existing field name and additionally vector itself appends event fields which are generally not underscore prefixed.
- Notably the parsing of field names is strict- if an invalid character is detected in the field name (eg "_foo%bar"), we don't just drop the field from the event, we error the encoding. This is to make it more obvious to users and to be consistent with the GELF decoder.

### CODE REVIEW

File: lib/codecs/src/encoding/format/gelf.rs
Comment (spencergilbert) [inline lib/codecs/src/encoding/format/gelf.rs:73]: Want to make sure we handle this properly before merging!

File: lib/codecs/src/encoding/format/gelf.rs
Comment (spencergilbert) [inline lib/codecs/src/encoding/format/gelf.rs:61]: We normally do regexes in a `once_cell`, is it worth doing that here as well?

File: lib/codecs/src/encoding/format/gelf.rs
Comment (spencergilbert) [inline lib/codecs/src/encoding/format/gelf.rs:99]: This reads awkwardly to me, I suspect there's a better way of modeling it...

File: lib/codecs/src/encoding/format/gelf.rs
Comment (spencergilbert) [inline lib/codecs/src/encoding/format/gelf.rs:80]: The multiple usages of this error message makes me think it should get pulled out into InternalEvents and shared? Perhaps the type/contents errors below as well.

File: lib/codecs/src/encoding/format/gelf.rs
Comment (spencergilbert) [inline lib/codecs/src/encoding/format/gelf.rs:173]: Is this still used? Or is it considered "conformed" when we prepend `_`

File: lib/codecs/src/encoding/format/gelf.rs
Comment (spencergilbert) [inline lib/codecs/src/encoding/format/gelf.rs:257]: I don't _love_ the name, I think we should unify on sanitize or conform (or something else) and then the function could be `gelf_serializing_conformable`

File: lib/codecs/src/encoding/format/gelf.rs
Comment (neuronull) [reply lib/codecs/src/encoding/format/gelf.rs:73]: yes... looking into it as we speak!

File: lib/codecs/src/encoding/format/gelf.rs
Comment (neuronull) [reply lib/codecs/src/encoding/format/gelf.rs:173]: good catch- before storing conformed in the struct, I was returning it. Need to update that comment.

File: lib/codecs/src/encoding/format/gelf.rs
Comment (neuronull) [reply lib/codecs/src/encoding/format/gelf.rs:257]: yes, good call. that name is remnant from the eradicated config option `sanitize`

File: lib/codecs/src/encoding/format/gelf.rs
Comment (spencergilbert) [reply lib/codecs/src/encoding/format/gelf.rs:73]: 🤦 I meant to block merging with my review too, but I clicked the wrong thing - whoops

File: lib/codecs/src/encoding/format/gelf.rs
Comment (neuronull) [reply lib/codecs/src/encoding/format/gelf.rs]: done

File: lib/codecs/src/encoding/format/gelf.rs
Comment (neuronull) [reply lib/codecs/src/encoding/format/gelf.rs]: took a stab, lmk if its not better

File: lib/codecs/src/encoding/format/gelf.rs
Comment (neuronull) [reply lib/codecs/src/encoding/format/gelf.rs]: Made an attempt at this as well.

File: lib/codecs/src/internal_events/encoder.rs
Comment (spencergilbert) [inline lib/codecs/src/internal_events/encoder.rs:11]: (thinking out loud) I don't know if it's easily accessible, but these could be more generic if the encoder type was passed in as a tag - rather than the as the struct name

File: lib/codecs/src/encoding/format/gelf.rs
Comment (spencergilbert) [inline lib/codecs/src/encoding/format/gelf.rs:83]: Do we know, in theory, when this could panic? When a metric or trace gets passed to the encoder? Could we wire up a configuration that actually triggers this, I'd think you'd need a sink that supports non-log events (as to not drop them) plus support for gelf encoding. Can `console` GELF encode?

Comment (spencergilbert) [reply lib/codecs/src/encoding/format/gelf.rs:83]: 👍 just trying to understand when it might panic - didn't notice the supports -> to_json relationship

Comment (neuronull) [reply lib/codecs/src/encoding/format/gelf.rs:83]: I'll add a brief comment to that effect for this function, because yeah you kind of have to go to the definition of the Trait to see how it works.

File: lib/codecs/src/internal_events/encoder.rs
Comment (neuronull) [reply lib/codecs/src/internal_events/encoder.rs:11]: I like the idea. Upon inspecting the Serializer module and thinking about it, I think the most straight forward way to accomplish that would just to pass it in as a string. Initially I thought, maybe we could put it at the Serializer level in mod.rs , but I don't think that would add much benefit because these internal events will be issued by the specific encoders themselves. So perhaps just defining a static string in the specific encoder file, and passing that in as a tag to the internal event construction would suffice?

Something like:

```Rust
static ENCODER_TYPE_GELF_STR : &str = "GELF";
```
...
```Rust
emit(SerializerFailedMissingField { field, message: MISSING_FIELD_STR, encoder_type: ENCODER_TYPE_GELF_STR });
```

Comment (spencergilbert) [reply lib/codecs/src/internal_events/encoder.rs:11]: Maybe here is an option as well? https://github.com/vectordotdev/vector/blob/master/src/internal_events/prelude.rs

**Review by spencergilbert — APPROVED** [review]
👍 but would like to get a review from @pablosichert in particular

File: lib/codecs/src/encoding/format/gelf.rs
Comment (jszwedko) [inline lib/codecs/src/encoding/format/gelf.rs:101]: I wonder if version is something that the encoder should set since we are controlling the GELF encoding here?

File: lib/codecs/src/encoding/format/gelf.rs
Comment (jszwedko) [inline lib/codecs/src/encoding/format/gelf.rs:231]: I think a better pattern here would be to have the function return the transformed event rather than having `validate_event_is_gelf` have a side-effect (setting `self.conformed_log`) that is checked later. We could call it something like `to_gelf_event`. I generally think this makes the control flow easier to follow.

File: lib/codecs/src/encoding/format/gelf.rs
Comment (jszwedko) [inline lib/codecs/src/encoding/format/gelf.rs:147]: Any reason if/else was preferred here over `match`?

**Review by jszwedko — COMMENTED** [review]
Looks good! Just a few notes below.

File: lib/codecs/src/encoding/format/gelf.rs
Comment (neuronull) [reply lib/codecs/src/encoding/format/gelf.rs:101]: That is a good point, for cases where events received were not already GELF.

File: lib/codecs/src/encoding/format/gelf.rs
Comment (neuronull) [reply lib/codecs/src/encoding/format/gelf.rs:231]: I'll look at that approach again. I had that initially but given I have helper methods that can also create the conformed_log , I was placing the conformed_log as a member of the struct to avoid doing the copy unless we really needed to.

File: lib/codecs/src/encoding/format/gelf.rs
Comment (neuronull) [reply lib/codecs/src/encoding/format/gelf.rs:147]: Nope! `match` does look cleaner 👍

**Comment by pablosichert** [comment]
@neuronull I don't know what exactly happened in this PR, but it looks like there was a bad merge with `origin/master`. Happy to give it another review after it is resolved 😄

**Comment by neuronull** [comment]
> @neuronull I don't know what exactly happened in this PR, but it looks like there was a bad merge with `origin/master`. Happy to give it another review after it is resolved 😄

Thanks and yeah I'm not sure either exactly - I should have waited for auto rebase from github maybe ?
I will try to resolve it today.

**Comment by neuronull** [comment]
> > @neuronull I don't know what exactly happened in this PR, but it looks like there was a bad merge with `origin/master`. Happy to give it another review after it is resolved 😄
> 
> Thanks and yeah I'm not sure either exactly - I should have waited for auto rebase from github maybe ? I will try to resolve it today.

Hey @pablosichert , I believe I sorted out the bad merge ! Let me know if not, thanks!

File: lib/codecs/src/encoding/format/gelf.rs
Comment (pablosichert) [inline lib/codecs/src/encoding/format/gelf.rs:100]: Could you elaborate a bit why this indirection is necessary?

File: lib/codecs/src/encoding/format/gelf.rs
Comment (neuronull) [reply lib/codecs/src/encoding/format/gelf.rs:231]: Perhaps I need to improve documentation here- the `to_gelf_event()` returns Err if we can't format to GELF. The Some(LogEvent) is returned if we applied any of the minor conformations that are possible. So in the else clause here, it means the buffer was already in valid GELF with no modifications needed.

File: lib/codecs/src/encoding/format/gelf.rs
Comment (neuronull) [reply lib/codecs/src/encoding/format/gelf.rs:74]: Oh... I think I made this mistake because the other Serializers that supported this were already encoded in json.

It seems then that perhaps the return value of this function should be:

```Rust
Result<serde_json::Value, vector_core::Error>
```

? Otherwise, would have to construct a serde_json Error if the encoding fails.

Comment (pablosichert) [reply lib/codecs/src/encoding/format/gelf.rs:74]: Makes sense, feel free to change the signatures where fit!

(You can construct a `serde_json::Error` by calling [`serde_json::Error::custom`](https://docs.serde.rs/serde/de/trait.Error.html#tymethod.custom), but I agree that it doesn't necessarily fit well here.)

Comment (neuronull) [reply lib/codecs/src/encoding/format/gelf.rs:74]: Will do- I see now the comment in the method is pretty clear, I just overlooked it 😬  "Encode event ..."

File: lib/codecs/src/encoding/format/gelf.rs
Comment (neuronull) [reply lib/codecs/src/encoding/format/gelf.rs]: Attempted to address in latest commit.

File: lib/codecs/src/encoding/format/gelf.rs
Comment (neuronull) [reply lib/codecs/src/encoding/format/gelf.rs]: Attempted to address in latest commit.

File: lib/codecs/src/encoding/format/gelf.rs
Comment (neuronull) [reply lib/codecs/src/encoding/format/gelf.rs:124]: Attempted to address in latest commit.

File: lib/codecs/src/encoding/format/gelf.rs
Comment (neuronull) [reply lib/codecs/src/encoding/format/gelf.rs:261]: Attempted to address in latest commit.

File: lib/codecs/src/encoding/format/gelf.rs
Comment (jszwedko) [inline lib/codecs/src/encoding/format/gelf.rs:232]: Apologies for harping on this, but I still find this interface a bit confusing 😅  It's definitely improved from storing the mutated log event on `self` but, what I would expect is for this function to take a `LogEvent` and then return that `LogEvent` (potentially modified). That is the signature would be:

```
    fn to_gelf_event(&self, log: LogEvent) -> vector_core::Result<LogEvent> {
```

Passing the owned log event around would avoid the need to have a separate `conformed_log` argument for the helper functions. The below would then become:

```
let log = self.coerce_required_fields(log)?;
let log = self.coerce_field_names_and_values(log)?;
```

(or you could chain them together with `and_then`: `self.coerce_required_fields(log).and_then(|log| self.coerce_field_names_and_values(log)`)

Does that make sense? I'm curious what you think.

File: lib/codecs/src/encoding/format/gelf.rs
Comment (jszwedko) [inline lib/codecs/src/encoding/format/gelf.rs:99]: Not a big deal, but this function, and some of the below, don't seem to need `self`. Maybe they could be pulled out as private helper functions outside of this `impl` block?

**Review by jszwedko — COMMENTED** [review]
Coming along! Just a couple more comments.

File: lib/codecs/src/encoding/format/gelf.rs
Comment (spencergilbert) [reply lib/codecs/src/encoding/format/gelf.rs:99]: I'm surprised there isn't a clippy ding for these situations?

File: lib/codecs/src/encoding/format/gelf.rs
Comment (neuronull) [reply lib/codecs/src/encoding/format/gelf.rs:232]: Harping is good and appreciated! - as I'm still developing my Rust skills.

I believe what led me to this, is I was trying to avoid was making a copy of the incoming event from `encode()` every time, because we don't always have to do that.

The approach you suggest does make sense, and I'm happy to do it, was just trying to optimize with less copies. But maybe there is a way to achieve that without copying... I'll take another look.

File: lib/codecs/src/encoding/format/gelf.rs
Comment (neuronull) [reply lib/codecs/src/encoding/format/gelf.rs:99]: Ah good catch , this happened to me previously as well . I was also surprised clippy didn't warn about it.

Comment (neuronull) [reply lib/codecs/src/encoding/format/gelf.rs:99]: Because `GelfSerializer` has no member variables, it seems most of those functions don't need to be in the `impl` block ... At one point I had a member variable there so it made sense. But even without it , I think I was just keeping them in the `impl` block for organization sake. I'll pull them out though.

File: lib/codecs/src/encoding/format/gelf.rs
Comment (jszwedko) [reply lib/codecs/src/encoding/format/gelf.rs:232]: Totally. That makes sense to avoid the clone. I believe my suggestion here to take the owned event should avoid the clone as well by just (possibly) mutating it and then returning it back. I could be missing something though.

Comment (neuronull) [reply lib/codecs/src/encoding/format/gelf.rs:232]: Ok I'll take a closer look at it. I'm doing an MVP test here and not getting it to work due to ownership rules but I might also be missing something.

Comment (neuronull) [reply lib/codecs/src/encoding/format/gelf.rs:232]: Ok I've given this another look and, it seems if I use the function signature you outlined:

>     fn to_gelf_event(&self, log: LogEvent) -> vector_core::Result<LogEvent> {

, The only way to make that work is to call it by:

```Rust
to_gelf_event(log.to_owned())
```

, and unless I'm mistaken, that is doing a copy.

Comment (jszwedko) [reply lib/codecs/src/encoding/format/gelf.rs:232]: I think that is because you are calling `.as_log()` above at the call site, which turns an `Event` into an `&LogEvent`. Rather than `.as_log()` you can use `.into_log()` to convert the `Event` into an owned `LogEvent`. It looks like all of the places where `to_gelf_event` is called receive an owned `Event` so it seemingly should be possible.

Comment (neuronull) [reply lib/codecs/src/encoding/format/gelf.rs:232]: Started to look into this- the `.into_log()` is enlightening :)

If `into_log()` converts to an owned LogEvent , then is there any reason we couldn't do:

```Rust
let mut log = event.into_log();
to_gelf_event(&mut log)?;
```

(and the other helper methods receive a `log : &mut LogEvent` , as parameter ?

Comment (jszwedko) [reply lib/codecs/src/encoding/format/gelf.rs:232]: That should work too. My personal preference is still for passing owned values in and returning owned values out rather than passing in a mutable reference `&mut` but I don't feel super strongly about it. Either way is a significant improvement over the current state in my opinion.

Comment (neuronull) [reply lib/codecs/src/encoding/format/gelf.rs]: I think I resolved it in the latest commit.
Had to extract the modification of the additional field keys missing underscore prefix, as was not able to modify the log event while already accessing the map.

File: lib/codecs/src/encoding/format/gelf.rs
Comment (jszwedko) [inline lib/codecs/src/encoding/format/gelf.rs:222]: Not a big deal, but could the missing prefix handling also be `and_then()`d above?

**Review by jszwedko — APPROVED** [review]
Nice! Looks great! I'd still like to get @pablosichert's review on this but it looks good to me.

File: lib/codecs/src/encoding/mod.rs
Comment (neuronull) [reply lib/codecs/src/encoding/mod.rs:353]: (In case you missed it), this was kind of discussed with Pablo above in his most recent review, but there may have been undesirable outcomes from changing the fn signature.

File: lib/codecs/src/encoding/format/gelf.rs
Comment (jszwedko) [inline lib/codecs/src/encoding/format/gelf.rs:226]:
```suggestion
    let mut missing_prefix = vec![];
    coerce_required_fields(log).and_then(|log| {
        coerce_field_names_and_values(log, &mut missing_prefix).map(|mut log| {
            // rename additional fields that were flagged as missing the underscore prefix
            for field in missing_prefix {
                log.rename_key(path!(field.as_str()), &*format!("_{}", &field));
            }
            log
        })
    })
```

Now we can clean this up even more 😄 

You could take this even further by having `coerce_field_names_and_values` actually return the missing prefixes, rather than taking an "out" parameter. That is, the signature would be:

```rust
fn coerce_field_names_and_values(
    mut log: LogEvent,
) -> vector_core::Result<(LogEvent, Vec<String>)> {
```

Then the above could just be:

```rust
    coerce_required_fields(log).and_then(|log| {
        coerce_field_names_and_values(log)
   }).and_then(|(mut log, missing_prefix)| {
            // rename additional fields that were flagged as missing the underscore prefix
            for field in missing_prefix {
                log.rename_key(path!(field.as_str()), &*format!("_{}", &field));
            }
            log
    })
```

Comment (neuronull) [reply lib/codecs/src/encoding/format/gelf.rs:226]: Oh excellent, yes that is better.

File: lib/codecs/src/encoding/mod.rs
Comment (pablosichert) [reply lib/codecs/src/encoding/mod.rs:353]: This has been discussed / suggested in https://github.com/vectordotdev/vector/pull/13333#discussion_r916969968.

Since the `to_json_value` is supposed to encode the value by its format first, it makes sense to expand the error type for more generic errors that can occur either while encoding or converting to `JSON`

File: lib/codecs/src/encoding/format/gelf.rs
Comment (pablosichert) [inline lib/codecs/src/encoding/format/gelf.rs:219]: Interesting, is the `&*` strictly needed here?

File: lib/codecs/src/encoding/mod.rs
Comment (pablosichert) [inline lib/codecs/src/encoding/mod.rs:337]: I don't think this change is necessary.

The `GelfSerializer` already informed the sink that it only accepts `DataType::Log`, so this will always be true

**Review by pablosichert — COMMENTED** [review]
@neuronull this PR already looks very good!

I have a last round of (mostly pedantic) comments:

File: lib/codecs/src/encoding/format/gelf.rs
Comment (pablosichert) [reply lib/codecs/src/encoding/format/gelf.rs]: This suggestion has not been applied yet, or did I miss it?

One thing to add to @jszwedko's suggestion that clippy will probably complain about:

```rust
.and_then(|log| {
    coerce_field_names_and_values(log)
})
```

can be just

```rust
.and_then(coerce_field_names_and_values)
```

File: lib/codecs/src/encoding/format/gelf.rs
Comment (neuronull) [reply lib/codecs/src/encoding/format/gelf.rs]: Good observation- I was able to replace that by calling `as_str()` instead- much more readable.

File: lib/codecs/src/encoding/format/gelf.rs
Comment (neuronull) [reply lib/codecs/src/encoding/format/gelf.rs]: clippy complained about the `and_then` returning Ok(()) being able to be replaced by `map()` , so that is what I went with here  :)

**Review by pablosichert — APPROVED** [review]
Awesome, nice work here @neuronull 🎉

**Pull request merged**
